core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java

```diff
@@ -157,170 +155,170 @@ public DefaultMessageBuilder(org.mule.runtime.api.message.Message message) {
 
   @Override
   public CollectionBuilder itemMediaType(MediaType mediaType) {
-    if (dataType instanceof DefaultCollectionDataType) {
-      dataType =
-          ((DataTypeBuilder.DataTypeCollectionTypeBuilder) DataType.builder(this.dataType)).itemMediaType(mediaType).build();
+    if (payload.getDataType() instanceof DefaultCollectionDataType) {
+      payload = new TypedValue(payload.getValue(),
+                               ((DataTypeBuilder.DataTypeCollectionTypeBuilder) builder(payload.getDataType()))
+                                   .itemMediaType(mediaType).build());
     } else {
       throw new IllegalStateException("Item MediaType cannot be set, because payload is not a collection");
     }
     return this;
   }
 
   @Override
-  public InternalMessage.Builder attributes(TypedValue<?> typedValue) {
-    this.attributes = typedValue.getValue();
-    this.attributesDataType = typedValue.getDataType();
+  public InternalMessage.CollectionBuilder attributes(TypedValue<?> attributes) {
```

Comment:
@elrodro83 can you give your take on this? At this point I'm inclined to say we should just have Builder if it gets too complicated. I'd expect CollectionBuilder to only be returned in the `collectionValue` methods, to force `itemMediaType` or something. Everywhere else it should be Builder.